packages/ui-kit/src/components/TimeRangePicker/TimeRangePicker.tsx

```diff
@@ -0,0 +1,454 @@
+import React from 'react'
+import { ClassNames, DateRange, DayPicker } from 'react-day-picker'
+import styles from 'react-day-picker/dist/style.module.css'
+import { useForwardedRefCallback } from '../../helpers'
+import { Button } from '../Button'
+import { Divider } from '../Divider'
+import { CalendarIcon } from '../Icons/Calendar'
+import { CloseIcon } from '../Icons/Close'
+import { Input } from '../Input'
+import { Select } from '../Select'
+import { Option } from '../Select/Option'
+import { useSetupTheme } from '../ThemeProvider'
+import { Typography } from '../Typography'
+import { DateTimeField } from './DateTimeField'
+import {
+  CUSTOM_DATE_RANGE,
+  CUSTOM_RANGE_FORMAT,
+  DATE_FORMAT,
+  defaultOptions,
+  FROM_DATE_UNTIL_NOW,
+  lastNOptions
+} from './TimeRangePicker.const'
+import componentStyles from './TimeRangePicker.module.scss'
+import { DateRangeOptionsProps, TimeRangePickerProps } from './TimeRangePicker.types'
+
+const formatDateTime = (value: Date | undefined, valueFormat: string) => (value ? format(value, valueFormat) : '')
+
+export const TimeRangePicker = React.forwardRef<HTMLDivElement, TimeRangePickerProps>(
+  (
+    {
+      baseTheme,
+      disableDateUntilNow = false,
+      disableCustomRange = false,
+      disableCustomRelative = false,
+      disableOptions = false,
+      disabled = false,
+      options: optionsProp,
+      defaultValue: defaultValueProp,
+      value,
+      onChange
+    },
+    forwardedRef
+  ) => {
+    const { componentContainer, setRef } = useForwardedRefCallback(forwardedRef)
+    const { theme } = useSetupTheme({
+      baseTheme,
+      componentContainer
+    })
+    const options = React.useMemo(() => (optionsProp ? optionsProp(defaultOptions) : defaultOptions), [optionsProp])
+    const [defaultValue] = React.useState(defaultValueProp)
+
+    const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null)
+    const [selectOpen, setSelectOpen] = React.useState(false)
+
+    const [lastN, setLastN] = React.useState(30)
+    const [lastNOption, setLastNOption] = React.useState('days')
+
+    const [selectedOptionLabel, setSelectedOptionLabel] = React.useState<string | null>(null)
+    const [selectedOption, setSelectedOption] = React.useState<DateRangeOptionsProps | undefined>(defaultValue)
+    const [datepickerRange, setDatepickerRange] = React.useState<DateRange | null>(null)
+    const [selectedRange, setSelectedRange] = React.useState<DateRange | null>(null)
+
+    const [datepickerMode, setDatepickerMode] = React.useState<typeof FROM_DATE_UNTIL_NOW | typeof CUSTOM_DATE_RANGE>(
+      FROM_DATE_UNTIL_NOW
+    )
+    const popupOpen = Boolean(anchorEl)
+
+    const timeoutRef = React.useRef<NodeJS.Timeout | null>(null)
+    const lastNRef = React.useRef(lastN)
+    const lastNOptionRef = React.useRef(lastNOption)
+
+    React.useEffect(() => {
+      lastNRef.current = lastN
+      lastNOptionRef.current = lastNOption
+    }, [lastN, lastNOption])
```

Comment:
What's the reason for constantly copying `lastN` and `lastNOption` from state to a ref? Why not using `lastN` and `lastNOption` directly?

Reply:
It allows storing mutable values that persist across renders without causing re-renders when they change. The whole thing is because of the `debouncedUpdate` logic.